File: implementation/scipamato/scipamato-core-entity/src/test/java/ch/difty/scipamato/core/auth/RoleTest.java
package ch.difty.scipamato.core.auth;

import static ch.difty.scipamato.core.auth.Role.*;
import static org.assertj.core.api.Assertions.*;

import org.junit.Test;

public class RoleTest {

    @Test
    public void hasAllValues() {
        assertThat(Role.values()).containsExactly(ADMIN, USER, VIEWER);
    }

    @Test
    public void assertIds() {
        assertThat(extractProperty("id").from(Role.values())).containsExactly(1, 2, 3);
    }

    @Test
    public void assertKeys() {
        assertThat(extractProperty("key").from(Role.values())).containsExactly(Roles.ADMIN, Roles.USER, Roles.VIEWER);
    }

    @Test
    public void assertDescriptions() {
        assertThat(extractProperty("description").from(Role.values())).containsExactly("System Administration",
            "Main SciPaMaTo Users", "Read-only Viewer");
    }

    @Test
    public void assertToString() {
        assertThat(ADMIN.toString()).isEqualTo("ROLE_ADMIN");
        assertThat(USER.toString()).isEqualTo("ROLE_USER");
        assertThat(VIEWER.toString()).isEqualTo("ROLE_VIEWER");
    }

    @Test
    public void of_withExistingId() {
        assertThat(Role.of(1)).isEqualTo(ADMIN);
    }

    @Test
    public void of_withNotExistingId_throws() {
        try {
            Role.of(0);
            fail("Should have thrown exception");
        } catch (Exception ex) {
            assertThat(ex)
                .isInstanceOf(IllegalArgumentException.class)
                .hasMessage("No matching type for id 0");
        }
    }

    @Test
    public void of_withNullId_returnsNull() {
        assertThat(Role.of(null)).isNull();
    }

}

File: core/core-entity/src/test/kotlin/ch/difty/scipamato/core/auth/RoleTest.kt
package ch.difty.scipamato.core.auth

import ch.difty.scipamato.core.auth.Role.ADMIN
import ch.difty.scipamato.core.auth.Role.USER
import ch.difty.scipamato.core.auth.Role.VIEWER
import ch.difty.scipamato.core.auth.Role.of
import org.amshove.kluent.invoking
import org.amshove.kluent.shouldBeEqualTo
import org.amshove.kluent.shouldBeNull
import org.amshove.kluent.shouldContainAll
import org.amshove.kluent.shouldContainSame
import org.amshove.kluent.shouldThrow
import org.amshove.kluent.withMessage
import org.junit.jupiter.api.Test

internal class RoleTest {

    @Test
    fun hasAllValues() {
        Role.entries shouldContainAll listOf(ADMIN, USER, VIEWER)
    }

    @Test
    fun assertIds() {
        Role.entries.map { it.id } shouldContainAll listOf(1, 2, 3)
    }

    @Test
    fun assertKeys() {
        Role.entries.map { it.key } shouldContainAll listOf(Roles.ADMIN, Roles.USER, Roles.VIEWER)
    }

    @Test
    fun assertDescriptions() {
        Role.entries.map { it.description } shouldContainSame
            listOf("System Administration", "Main SciPaMaTo Users", "Read-only Viewer")
    }

    @Test
    fun assertToString() {
        ADMIN.toString() shouldBeEqualTo "ROLE_ADMIN"
        USER.toString() shouldBeEqualTo "ROLE_USER"
        VIEWER.toString() shouldBeEqualTo "ROLE_VIEWER"
    }

    @Test
    fun of_withExistingId() {
        of(1) shouldBeEqualTo ADMIN
    }

    @Test
    fun of_withNotExistingId_throws() {
        invoking { of(0) } shouldThrow IllegalArgumentException::class withMessage "No matching type for id 0"
    }

    @Test
    fun of_withNullId_returnsNull() {
        of(null).shouldBeNull()
    }
}
